File: f77/build09/kmtc.f
C
        Subroutine Kinematics_Data
C
C       This subroutine sets the Default data for the kinematics routine.
C
C       Inputs:
C           None.
C
C     Outputs:
C       Acc_Gravity   - Acceleration due to gravity [m/sec**2]
C       Psi_b_IC_dg   - Initial Euler angles [deg]
C       Theta_b_IC_dg
C       Phi_b_IC_dg
C       P_b_IC_dg     - Initial missile rotation velocity [deg/sec]
C       Q_b_IC_dg
C       R_b_IC_dg
C       Xd_bi_i_IC    - Initial missile velocity WRT ICS [m/sec]
C       Yd_bi_i_IC
C       Zd_bi_i_IC
C       X_bi_i_IC     - Initial missile position [m]
C       Y_bi_i_IC
C       Z_bi_i_IC
C
C     Internal variables and constants:
C       None.
C
C     Declare global common and assign variable locations
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array(536), Psi_b_IC_dg)
      Equivalence (Real_Array(537), Theta_b_IC_dg)
      Equivalence (Real_Array(538), Phi_b_IC_dg)
      Equivalence (Real_Array(539), P_b_IC_dg)
      Equivalence (Real_Array(540), Q_b_IC_dg)
      Equivalence (Real_Array(541), R_b_IC_dg)
      Equivalence (Real_Array(542), Xd_bi_i_IC)
      Equivalence (Real_Array(543), Yd_bi_i_IC)
      Equivalence (Real_Array(544), Zd_bi_i_IC)
      Equivalence (Real_Array(545), X_bi_i_IC)
      Equivalence (Real_Array(546), Y_bi_i_IC)
      Equivalence (Real_Array(547), Z_bi_i_IC)
      Equivalence (Real_Array(548), Acc_Gravity)
C
C     Begin default data definitions:
C
C     Variables for determining initial attitude
C
      Psi_b_IC_dg   = 0.0e0
      Theta_b_IC_dg = 0.0e0
      Phi_b_IC_dg   = 0.0e0
C
C     Variables for determining initial angular velocity
C
      P_b_IC_dg = 0.0e0
      Q_b_IC_dg = 0.0e0
      R_b_IC_dg = 0.0e0
C
C     Variables for determining initial translational velocity of the BCS w.r.t.
C     the ICS
C
      Xd_bi_i_IC = 100.0e0
      Yd_bi_i_IC =   0.0e0
      Zd_bi_i_IC =   0.0e0
C
C     Variables for determining initial translational position of the BCS w.r.t.
C     the ICS
C
      X_bi_i_IC =    0.0e0
      Y_bi_i_IC =    0.0e0
      Z_bi_i_IC = -100.0e0
C
C     Acceleration due to gravity
C
      Acc_Gravity = 9.88e0
C
      Return
      End

      Subroutine Kinematics_Init
C
C     Missile kinematics model initialization; initializes quaternions from
C     initial missile Euler angles.
C
C     Inputs:
C       Psi_b_IC_dg   - Initial Euler angles [deg]
C       Theta_b_IC_dg
C       Phi_b_IC_dg
C       P_b_IC_dg     - Initial missile rotational velocity [deg/sec]
C       Q_b_IC_dg
C       R_b_IC_dg
C       Xd_bi_i_IC    - Initial missile velocity WRT ICS [m/sec]
C       Yd_bi_i_IC
C       Zd_bi_i_IC
C       X_bi_i_IC     - Initial missile position [m]
C       Y_bi_i_IC
C       Z_bi_i_IC
C
C     Outputs:
C       P_b     - Missile inertial angular velocity [rad/sec]
C       Q_b
C       R_b
C       Q0_b    - ICS to BCS quaternion [Real]
C       Q1_b
C       Q2_b
C       Q3_b
C       X_bi_i  - Missile terminal position WRT ICS [m]
C       Y_bi_i
C       Z_bi_i
C       Xd_bi_i - Missile velocity in ICS [m/sec]
C       Yd_bi_i
C       Zd_bi_i
C
C     Internal variables and constants:
C       CPsiO2   - Cosines of half Euler angles [Real]
C       CThetaO2
C       CPhiO2
C       RDTODG   - Radians to degrees conversion factor [deg/rad]
C       SPsiO2   - Sines of half Euler angles [Real]
C       SThetaO2
C       SPhiO2
C
C     Declare global common and assign variable locations
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array( 51), RDTODG)
      Equivalence (Real_Array(500), X_bi_i)
      Equivalence (Real_Array(501), Xd_bi_i)
      Equivalence (Real_Array(503), Y_bi_i)
      Equivalence (Real_Array(504), Yd_bi_i)
      Equivalence (Real_Array(506), Z_bi_i)
      Equivalence (Real_Array(507), Zd_bi_i)
      Equivalence (Real_Array(509), P_b)
      Equivalence (Real_Array(511), Q_b)
      Equivalence (Real_Array(513), R_b)
      Equivalence (Real_Array(515), Psi_b)
      Equivalence (Real_Array(516), Theta_b)
      Equivalence (Real_Array(517), Phi_b)
      Equivalence (Real_Array(518), Q0_b)
      Equivalence (Real_Array(520), Q1_b)
      Equivalence (Real_Array(522), Q2_b)
      Equivalence (Real_Array(524), Q3_b)
      Equivalence (Real_Array(536), Psi_b_IC_dg)
      Equivalence (Real_Array(537), Theta_b_IC_dg)
      Equivalence (Real_Array(538), Phi_b_IC_dg)
      Equivalence (Real_Array(539), P_b_IC_dg)
      Equivalence (Real_Array(540), Q_b_IC_dg)
      Equivalence (Real_Array(541), R_b_IC_dg)
      Equivalence (Real_Array(542), Xd_bi_i_IC)
      Equivalence (Real_Array(543), Yd_bi_i_IC)
      Equivalence (Real_Array(544), Zd_bi_i_IC)
      Equivalence (Real_Array(545), X_bi_i_IC)
      Equivalence (Real_Array(546), Y_bi_i_IC)
      Equivalence (Real_Array(547), Z_bi_i_IC)
C
C     Assign state and state derivative pointers
C
      Call Define_Real_State(500, 501)
      Call Define_Real_State(501, 502)
      Call Define_Real_State(503, 504)
      Call Define_Real_State(504, 505)
      Call Define_Real_State(506, 507)
      Call Define_Real_State(507, 508)
      Call Define_Real_State(509, 510)
      Call Define_Real_State(511, 512)
      Call Define_Real_State(513, 514)
      Call Define_Real_State(518, 519)
      Call Define_Real_State(520, 521)
      Call Define_Real_State(522, 523)
      Call Define_Real_State(524, 525)
C
C     Begin math model initialization
C
C     Initial Euler angles: (BCS) = [ROLL][PITCH][YAW](ICS)
C
      Psi_b   = Psi_b_IC_dg/RDTODG
      Theta_b = Theta_b_IC_dg/RDTODG
      Phi_b   = Phi_b_IC_dg/RDTODG
C
C     Initial angular velocity of BCS w.r.t. ICS, expressed in BCS
C
      P_b = P_b_IC_dg/RDTODG
      Q_b = Q_b_IC_dg/RDTODG
      R_b = R_b_IC_dg/RDTODG
C
C     Initial translational position of the BCS w.r.t. the ICS
C
      X_bi_i = X_bi_i_IC
      Y_bi_i = Y_bi_i_IC
      Z_bi_i = Z_bi_i_IC
C
C     Initial translational velocity of the BCS w.r.t. to the ICS
C
      Xd_bi_i = Xd_bi_i_IC
      Yd_bi_i = Yd_bi_i_IC
      Zd_bi_i = Zd_bi_i_IC
C
C     Trig functions of half Euler angles
C
      CPsiO2   = Cos (0.5e0*Psi_b)
      CThetaO2 = Cos (0.5e0*Theta_b)
      CPhiO2   = Cos (0.5e0*Phi_b)
      SPsiO2   = Sin (0.5e0*Psi_b)
      SThetaO2 = Sin (0.5e0*Theta_b)
      SPhiO2   = Sin (0.5e0*Phi_b)
C
C     Initial quaternion values
C
      Q0_b = CPsiO2*CThetaO2*CPhiO2 + SPsiO2*SThetaO2*SPhiO2
      Q1_b = CPsiO2*CThetaO2*SPhiO2 - SPsiO2*SThetaO2*CPhiO2
      Q2_b = CPsiO2*SThetaO2*CPhiO2 + SPsiO2*CThetaO2*SPhiO2
      Q3_b = SPsiO2*CThetaO2*CPhiO2 - CPsiO2*SThetaO2*SPhiO2
C
      Return
      End

      Subroutine Kinematics
C
C     This module evaluates the translational and rotational derivatives which
C     describe the missile's motion WRT an earth fixed coordinate system for
C     a non-rotating earth.
C
C     Inputs:
C       AaeroX_bi_b - Accelerations due to aero in BCS [m/sec**2]
C       AaeroY_bi_b
C       AaeroZ_bi_b
C       Acc_Gravity - Acceleration due to gravity [m/sec**2]
C       P_b         - Missile inertial angular velocity [rad/sec]
C       Q_b
C       R_b
C       Q0_b        - ICS to BCS quaternion [Real]
C       Q1_b
C       Q2_b
C       Q3_b
C       Tibij       - ICS to BCS transformation matrix [Real]
C
C     Outputs:
C       Q0_b     - ICS to BCS quaternion [Real]
C       Q1_b
C       Q2_b
C       Q3_b
C       Qd0_b    - ICS to BCS quaternion derivatives [Real]
C       Qd1_b
C       Qd2_b
C       Qd3_b
C       Psi_b    - Missile Euler angles [rad]
C       Theta_b
C       Phi_b
C       Xdd_bi_i - Missile acceleration WRT ICS in ICS [m/sec**2]
C       Ydd_bi_i
C       Zdd_bi_i
C
C     Internal variables and constants:
C       Qmag     - Magnitude of ICS to BCS quaternion [Real]
C       RDTODG   - Radians to degrees conversion factor [deg/rad]
C       STheta_b - Sine of pitch Euler angle [Real]
C
C     Declare global common and assign variable locations
C
      Include '../driver/global.inc'
      Include '../driver/sysvars.inc'
      Equivalence (Real_Array( 51), RDTODG)
      Equivalence (Real_Array(100), AaeroX_bi_b)
      Equivalence (Real_Array(101), AaeroY_bi_b)
      Equivalence (Real_Array(102), AaeroZ_bi_b)
      Equivalence (Real_Array(502), Xdd_bi_i)
      Equivalence (Real_Array(505), Ydd_bi_i)
      Equivalence (Real_Array(508), Zdd_bi_i)
      Equivalence (Real_Array(509), P_b)
      Equivalence (Real_Array(511), Q_b)
      Equivalence (Real_Array(513), R_b)
      Equivalence (Real_Array(515), Psi_b)
      Equivalence (Real_Array(516), Theta_b)
      Equivalence (Real_Array(517), Phi_b)
      Equivalence (Real_Array(518), Q0_b)
      Equivalence (Real_Array(519), Q0d_b)
      Equivalence (Real_Array(520), Q1_b)
      Equivalence (Real_Array(521), Q1d_b)
      Equivalence (Real_Array(522), Q2_b)
      Equivalence (Real_Array(523), Q2d_b)
      Equivalence (Real_Array(524), Q3_b)
      Equivalence (Real_Array(525), Q3d_b)
      Equivalence (Real_Array(526), Tib11)
      Equivalence (Real_Array(527), Tib12)
      Equivalence (Real_Array(528), Tib13)
      Equivalence (Real_Array(529), Tib21)
      Equivalence (Real_Array(530), Tib22)
      Equivalence (Real_Array(531), Tib23)
      Equivalence (Real_Array(532), Tib31)
      Equivalence (Real_Array(533), Tib32)
      Equivalence (Real_Array(534), Tib33)
      Equivalence (Real_Array(548), Acc_Gravity)
C
C     Begin math model:
C
C     Evaluate Euler angles
C
      STheta_b = Limit (-Tib13, -1.0e0, 1.0e0)
      Theta_b  = Asin (STheta_b)
      Psi_b    = ArcTan (Tib12, Tib11)
      Phi_b    = ArcTan (Tib23, Tib33)
C
C     Missile velocity rate in ICS WRT ICS
C
      Xdd_bi_i = Tib11*AaeroX_bi_b +
     &           Tib21*AaeroY_bi_b +
     &           Tib31*AaeroZ_bi_b
      Ydd_bi_i = Tib12*AaeroX_bi_b +
     &           Tib22*AaeroY_bi_b +
     &           Tib32*AaeroZ_bi_b
      Zdd_bi_i = Tib13*AaeroX_bi_b +
     &           Tib23*AaeroY_bi_b +
     &           Tib33*AaeroZ_bi_b +
     &           Acc_Gravity
C
C     Quaternion constraint equation
C
      Qmag = Sqrt (Q0_b*Q0_b + Q1_b*Q1_b + Q2_b*Q2_b + Q3_b*Q3_b)
      Q0_b = Q0_b/Qmag
      Q1_b = Q1_b/Qmag
      Q2_b = Q2_b/Qmag
      Q3_b = Q3_b/Qmag
C
C     Quaternion derivative
C
      Q0d_b = -0.5e0*(Q1_b*P_b + Q2_b*Q_b + Q3_b*R_b)
      Q1d_b =  0.5e0*(Q0_b*P_b + Q2_b*R_b - Q3_b*Q_b)
      Q2d_b =  0.5e0*(Q0_b*Q_b + Q3_b*P_b - Q1_b*R_b)
      Q3d_b =  0.5e0*(Q0_b*R_b + Q1_b*Q_b - Q2_b*P_b)
C
      Return
      End

File: compcert/build09/kmtc.c
#include "kmtc.h"
#include "../driver/global.h"
#include "../driver/sysvars.h"
#include "../driver/deriv.h"
#include "../driver/defstate.h"
#include "../driver/limit.h"
#include <math.h>

#define rdtodg        real_array[ 51].r
#define aaerox_bi_b   real_array[100].r
#define aaeroy_bi_b   real_array[101].r
#define aaeroz_bi_b   real_array[102].r
#define x_bi_i        real_array[500].r
#define xd_bi_i       real_array[501].r
#define xdd_bi_i      real_array[502].r
#define y_bi_i        real_array[503].r
#define yd_bi_i       real_array[504].r
#define ydd_bi_i      real_array[505].r
#define z_bi_i        real_array[506].r
#define zd_bi_i       real_array[507].r
#define zdd_bi_i      real_array[508].r
#define p_b           real_array[509].r
#define q_b           real_array[511].r
#define r_b           real_array[513].r
#define psi_b         real_array[515].r
#define theta_b       real_array[516].r
#define phi_b         real_array[517].r
#define q0_b          real_array[518].r
#define q0d_b         real_array[519].r
#define q1_b          real_array[520].r
#define q1d_b         real_array[521].r
#define q2_b          real_array[522].r
#define q2d_b         real_array[523].r
#define q3_b          real_array[524].r
#define q3d_b         real_array[525].r
#define tib11         real_array[526].r
#define tib12         real_array[527].r
#define tib13         real_array[528].r
#define tib21         real_array[529].r
#define tib22         real_array[530].r
#define tib23         real_array[531].r
#define tib31         real_array[532].r
#define tib32         real_array[533].r
#define tib33         real_array[534].r
#define psi_b_ic_dg   real_array[536].r
#define theta_b_ic_dg real_array[537].r
#define phi_b_ic_dg   real_array[538].r
#define p_b_ic_dg     real_array[539].r
#define q_b_ic_dg     real_array[540].r
#define r_b_ic_dg     real_array[541].r
#define xd_bi_i_ic    real_array[542].r
#define yd_bi_i_ic    real_array[543].r
#define zd_bi_i_ic    real_array[544].r
#define x_bi_i_ic     real_array[545].r
#define y_bi_i_ic     real_array[546].r
#define z_bi_i_ic     real_array[547].r
#define acc_gravity   real_array[548].r

void kinematics_data() {
  // Outputs:
  //   Acc_Gravity   - Acceleration due to gravity [m/sec**2]
  //   Psi_b_IC_dg   - Initial Euler angles [deg]
  //   Theta_b_IC_dg
  //   Phi_b_IC_dg
  //   P_b_IC_dg     - Initial missile rotation velocity [deg/sec]
  //   Q_b_IC_dg
  //   R_b_IC_dg
  //   Xd_bi_i_IC    - Initial missile velocity WRT ICS [m/sec]
  //   Yd_bi_i_IC
  //   Zd_bi_i_IC
  //   X_bi_i_IC     - Initial missile position [m]
  //   Y_bi_i_IC
  //   Z_bi_i_IC
  // Begin default data definitions:
  // Variables for determining initial attitude
  psi_b_ic_dg = 0.0;
  theta_b_ic_dg = 0.0;
  phi_b_ic_dg = 0.0;
  // Variables for determining initial angular velocity
  p_b_ic_dg = 0.0;
  q_b_ic_dg = 0.0;
  r_b_ic_dg = 0.0;
  // Variables for determining initial translational velocity of the BCS w.r.t.
  // the ICS
  xd_bi_i_ic = 100.0;
  yd_bi_i_ic = 0.0;
  zd_bi_i_ic = 0.0;
  // Variables for determining initial translational position of the BCS w.r.t.
  // the ICS
  x_bi_i_ic = 0.0;
  y_bi_i_ic = 0.0;
  z_bi_i_ic = -100.0;
  // Acceleration due to gravity
  acc_gravity = 9.88;
}

void kinematics_init() {
  // Missile kinematics model initialization; initializes quaternions from
  // initial missile Euler angles.
  //
  // Inputs:
  //   Psi_b_IC_dg   - Initial Euler angles [deg]
  //   Theta_b_IC_dg
  //   Phi_b_IC_dg
  //   P_b_IC_dg     - Initial missile rotational velocity [deg/sec]
  //   Q_b_IC_dg
  //   R_b_IC_dg
  //   Xd_bi_i_IC    - Initial missile velocity WRT ICS [m/sec]
  //   Yd_bi_i_IC
  //   Zd_bi_i_IC
  //   X_bi_i_IC     - Initial missile position [m]
  //   Y_bi_i_IC
  //   Z_bi_i_IC
  //
  // Outputs:
  //   P_b     - Missile inertial angular velocity [rad/sec]
  //   Q_b
  //   R_b
  //   Q0_b    - ICS to BCS quaternion [Real]
  //   Q1_b
  //   Q2_b
  //   Q3_b
  //   X_bi_i  - Missile terminal position WRT ICS [m]
  //   Y_bi_i
  //   Z_bi_i
  //   Xd_bi_i - Missile velocity in ICS [m/sec]
  //   Yd_bi_i
  //   Zd_bi_i
  //
  // Internal variables and constants:
  //   CPsiO2   - Cosines of half Euler angles [Real]
  //   CThetaO2
  //   CPhiO2
  //   RDTODG   - Radians to degrees conversion factor [deg/rad]
  //   SPsiO2   - Sines of half Euler angles [Real]
  //   SThetaO2
  //   SPhiO2
  //
  // Declare global common and assign variable locations
  REAL cpsio2,cthetao2,cphio2,spsio2,sthetao2,sphio2;

  // Assign state and state derivative pointers
  define_real_state(500,501);
  define_real_state(501,502);
  define_real_state(503,504);
  define_real_state(504,505);
  define_real_state(506,507);
  define_real_state(507,508);
  define_real_state(509,510);
  define_real_state(511,512);
  define_real_state(513,514);
  define_real_state(518,519);
  define_real_state(520,521);
  define_real_state(522,523);
  define_real_state(524,525);
  // Begin math model initialization
  // Initial Euler angles: (BCS) = [ROLL][PITCH][YAW](ICS)
  psi_b = psi_b_ic_dg/rdtodg;
  theta_b = theta_b_ic_dg/rdtodg;
  phi_b = phi_b_ic_dg/rdtodg;
  // Initial angular velocity of BCS w.r.t. ICS, expressed in BCS
  p_b = p_b_ic_dg/rdtodg;
  q_b = q_b_ic_dg/rdtodg;
  r_b = r_b_ic_dg/rdtodg;
  // Initial translational position of the BCS w.r.t. the ICS
  x_bi_i = x_bi_i_ic;
  y_bi_i = y_bi_i_ic;
  z_bi_i = z_bi_i_ic;
  // Initial translational velocity of the BCS w.r.t. to the ICS
  xd_bi_i = xd_bi_i_ic;
  yd_bi_i = yd_bi_i_ic;
  zd_bi_i = zd_bi_i_ic;
  // Trig functions of half Euler angles
  cpsio2 = cos(0.5*psi_b);
  cthetao2 = cos(0.5*theta_b);
  cphio2 = cos(0.5*phi_b);
  spsio2 = sin(0.5*psi_b);
  sthetao2 = sin(0.5*theta_b);
  sphio2 = sin(0.5*phi_b);
  // Initial quaternion values
  q0_b = cpsio2*cthetao2*cphio2 + spsio2*sthetao2*sphio2;
  q1_b = cpsio2*cthetao2*sphio2 - spsio2*sthetao2*cphio2;
  q2_b = cpsio2*sthetao2*cphio2 + spsio2*cthetao2*sphio2;
  q3_b = spsio2*cthetao2*cphio2 - cpsio2*sthetao2*sphio2;
}

void kinematics() {
  // This module evaluates the translational and rotational derivatives which
  // describe the missile's motion WRT an earth fixed coordinate system for
  // a non-rotating earth.
  //
  // Inputs:
  //   AaeroX_bi_b - Accelerations due to aero in BCS [m/sec**2]
  //   AaeroY_bi_b
  //   AaeroZ_bi_b
  //   Acc_Gravity - Acceleration due to gravity [m/sec**2]
  //   P_b         - Missile inertial angular velocity [rad/sec]
  //   Q_b
  //   R_b
  //   Q0_b        - ICS to BCS quaternion [Real]
  //   Q1_b
  //   Q2_b
  //   Q3_b
  //   Tibij       - ICS to BCS transformation matrix [Real]
  //
  // Outputs:
  //   Q0_b     - ICS to BCS quaternion [Real]
  //   Q1_b
  //   Q2_b
  //   Q3_b
  //   Qd0_b    - ICS to BCS quaternion derivatives [Real]
  //   Qd1_b
  //   Qd2_b
  //   Qd3_b
  //   Psi_b    - Missile Euler angles [rad]
  //   Theta_b
  //   Phi_b
  //   Xdd_bi_i - Missile acceleration WRT ICS in ICS [m/sec**2]
  //   Ydd_bi_i
  //   Zdd_bi_i
  //
  // Internal variables and constants:
  //   Qmag     - Magnitude of ICS to BCS quaternion [Real]
  //   RDTODG   - Radians to degrees conversion factor [deg/rad]
  //   STheta_b - Sine of pitch Euler angle [Real]
  //
  // Declare global common and assign variable locations
  REAL stheta_b,qmag;
  // Begin math model:
  // Evaluate Euler angles
  stheta_b = limit(-tib13,-1.0,1.0);
  theta_b = asin(stheta_b);
  psi_b = atan2(tib12,tib11);
  phi_b = atan2(tib23,tib33);
  // Missile velocity rate in ICS WRT ICS
  xdd_bi_i = tib11*aaerox_bi_b + tib21*aaeroy_bi_b + tib31*aaeroz_bi_b;
  ydd_bi_i = tib12*aaerox_bi_b + tib22*aaeroy_bi_b + tib32*aaeroz_bi_b;
  zdd_bi_i = tib13*aaerox_bi_b + tib23*aaeroy_bi_b + tib33*aaeroz_bi_b + acc_gravity;
  // Quaternion constraint equation
  qmag = sqrt(q0_b*q0_b+q1_b*q1_b+q2_b*q2_b+q3_b*q3_b);
  q0_b = q0_b/qmag;
  q1_b = q1_b/qmag;
  q2_b = q2_b/qmag;
  q3_b = q3_b/qmag;
  // Quaternion derivative
  q0d_b = -0.5*(q1_b*p_b+q2_b*q_b+q3_b*r_b);
  q1d_b =  0.5*(q0_b*p_b+q2_b*r_b-q3_b*q_b);
  q2d_b =  0.5*(q0_b*q_b+q3_b*p_b-q1_b*r_b);
  q3d_b =  0.5*(q0_b*r_b+q1_b*q_b-q2_b*p_b);
}
